## Problem 24: Lexicographic Permutations

A permutation is an ordered arrangement of objects. For example, 3124 is one possible permutation of the digits 1, 2, 3 and 4. If all of the permutations are listed numerically or alphabetically, we call it lexicographic order. The lexicographic permutations of 0, 1 and 2 are:

$012\ \ \ 021\ \ \ 102\ \ \ 120\ \ \ 201\ \ \ 210$

What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9?

In [1]:
import itertools

In [2]:
''' 
Returns the nth iteration of an array of numbers.

eg:
>>> proper_divisors(24)
>>> [1, 2, 3, 4, 6, 8, 12]
'''

def nth_permutation(number_array : list, n_iteration : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Returns the nth iteration of an array of numbers.
        
        Use the itertools package to generate the permutations and select the nth iteration.
            
            Use the permutations function in itertools to generate the permutations: 
            https://docs.python.org/3/library/itertools.html#itertools.permutations
                Return successive r length permutations of elements in the iterable.
                itertools.permutations(iterable, r=None)
            
            Use the islice function in itertools to select the nth iteration:
            https://docs.python.org/3/library/itertools.html#itertools.islice
                Make an iterator that returns selected elements from the iterable.
        
        
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : list : number_array : The array of numbers to find the permutations for.
        : int : n_iteration : The number of the iteration of permutation to return.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : permutation : The nth iteration of the purmutations of the number array.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> proper_divisors(24)
        >>> [1, 2, 3, 4, 6, 8, 12]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number_array, list): raise ValueError('Please enter a list of integers for the number_array argument.')
    for number in number_array:
        if not isinstance(number, int): raise ValueError('Please enter a list of integers for the number_array argument.')
    if not isinstance(n_iteration, int): raise ValueError('Please enter an integer >0 for the n_iteration argument.')
    if n_iteration < 1: raise ValueError('Please enter an integer >0 for the n_iteration argument.')
    
    permutation_list = next(itertools.islice(itertools.permutations(number_array), n_iteration-1, None))
    
    permutation = ''.join(str(number) for number in permutation_list)
    
    return permutation

In [3]:
problem_nth_permutation = 1_000_000
permutation_start = [0,1,2,3,4,5,6,7,8,9]
solution = nth_permutation(permutation_start, problem_nth_permutation)

print(f'The {problem_nth_permutation:,}th Lexicographic permutation of the array {permutation_start} is {solution}.')

The 1,000,000th Lexicographic permutation of the array [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] is 2783915460.


### Problem 24 Solution:

The 1,000,000th Lexicographic permutation of the array [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] is 2783915460.